proxy/src/transport/tls/config.rs

```diff
@@ -110,27 +168,56 @@ impl CommonSettings {
             })
     }
 
-    /// Stream changes to the files described by this `CommonSettings`.
-    ///
-    /// This will poll the filesystem for changes to the files at the paths
-    /// described by this `CommonSettings` every `interval`, and attempt to
-    /// load a new `CommonConfig` from the files again after each change.
-    ///
-    /// The returned stream consists of each subsequent successfully loaded
-    /// `CommonSettings` after each change. If the settings could not be
-    /// reloaded (i.e., they were malformed), nothing is sent.
-    ///
-    /// TODO: On Linux, this should be replaced with an `inotify` watch when
-    ///       available.
-    pub fn stream_changes(self, interval: Duration)
-        -> impl Stream<Item = CommonConfig, Error = ()>
+    #[cfg(target_os = "linux")]
+    fn stream_changes_inotify(&self)
+        -> Result<impl Stream<Item = (), Error = ()>, Error>
     {
-        self.change_timestamps(interval)
-            .filter_map(move |_|
-                CommonConfig::load_from_disk(&self)
-                    .map_err(|e| warn!("error reloading TLS config: {:?}, falling back", e))
-                    .ok()
-            )
+        use inotify::{Inotify, WatchMask};
+        // NOTE: If we used a less broad watch mask, we could probably avoid
+        //       reloading the certs multiple times when k8s modifies a
+        //       ConfigMap (a multi-step process that we see as a series of
+        //       CREATE, MOVED_TO, MOVED_FROM, and DELETE events). However,
+        //       this may not catch changes if the files we're watching *don't*
+        //       live in a k8s ConfigMap, and I think false positives are
+        //       much less harmful here than false negatives.
+        let mask = WatchMask::CREATE
+                 | WatchMask::MODIFY
+                 | WatchMask::DELETE
+                 | WatchMask::MOVE
+                 ;
+        let mut inotify = Inotify::init().map_err(Error::InotifyInit)?;
+
+        for path in &self.paths() {
```

Comment:
I think we should map each path to the parent and then throw away duplicates. In particular, the end-entity certificate and private key will (usually) be in the same directory so there's no reason to watch that directory twice, IIRC.

Reply:
Also, we should add a note that we do such de-duping to minimize the chances that we trigger the inotify limit.

Reply:
> I thought de-duping here would add unnecessary complexity (and, depending on how we do it, allocations). However, it would also avoid making some additional syscalls, so if you think that's valuable, I'd be happy to add it.

I'm not sure if to duplicate is counted against the inotify limit or not. It's not a big deal but I think it would be clearer if we somehow acknowledge that we understand there will be dupes and what we expect to happen in that case.

Reply:
Okay, will do.